### PR TITLE
Fix docstring to clarify logits usage for multiclass case

### LINKED ISSUE #50378: Categorical logits argument is treated as log probabilities
## 🐛 Bug

The `logit` argument to `torch.distributions.Categorical` is not treated as logits, rather as log probabilities

## To Reproduce

```kj
import torch as t
from torch.distributions import Categorical
probs = t.tensor([.3, .3, .399, .001])
y_test = t.tensor([0, 1, 2, 1, 2, 2, 3])
print(t.exp(Categorical(probs=probs).log_prob(y_test)))
print(t.exp(Categorical(logits=t.log(probs)).log_prob(y_test)))
print(t.exp(Categorical(logits=t.logit(probs)).log_prob(y_test)))
```

produces
```
tensor([0.3000, 0.3000, 0.3990, 0.3000, 0.3990, 0.3990, 0.0010])
tensor([0.3000, 0.3000, 0.3990, 0.3000, 0.3990, 0.3990, 0.0010])
tensor([0.2816, 0.2816, 0.4362, 0.2816, 0.4362, 0.4362, 0.0007])
```

## Expected behavior

I expect the last line to be the same as the first and the second line to be different from the first. Log probabilities are not logits.

```
logit(p) = log(p / (1 - p)) = log(p) - log(1 - p) != log(p)
```

## Environment

```
Collecting environment information...
PyTorch version: 1.7.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 10.14.6 (x86_64)
GCC version: Could not collect
Clang version: Could not collect
CMake version: version 3.19.2

Python version: 3.7 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] botorch==0.3.1
[pip3] gpytorch==1.2.0
[pip3] numpy==1.17.3
[pip3] torch==1.7.1
[pip3] torch-struct==0.4
[pip3] torchvision==0.8.2
[conda] Could not collect
```

cc @fritzo @neerajprad @alicanb @vishwakftw @nikitaved

### PR DESCRIPTION
**Pull request by neerajprad**
Fixes #50378.

Additionally, this has some minor fixes:
 - [x] Fix mean for half-cauchy to return `inf` instead of `nan`.
 - [x] Fix constraints/support for the relaxed categorical distribution.

### CODE REVIEW

File: torch/distributions/half_cauchy.py [hunk 1]
```diff
@@ -43,7 +43,7 @@ def scale(self):
 
     @property
     def mean(self):
-        return self.base_dist.mean
+        return torch.full(self._extended_shape(), math.inf, dtype=self.scale.dtype, device=self.scale.device)
```
Comment (neerajprad) [inline torch/distributions/half_cauchy.py:46]: @fehiepsi - It turns out that I had not taken a close look and the stats for half-normal/half-cauchy look fine as is. I still made this modification to return `inf` rather than `nan` for the half-cauchy, but let me know if you had any other correction in mind.

Comment (fehiepsi) [reply torch/distributions/half_cauchy.py:46]: Thanks, @neerajprad! Looks good to me. You'll need to fix `variance` too.

Comment (neerajprad) [reply torch/distributions/half_cauchy.py:46]: The variance for the base dist should already be `inf` so I left it as is.

File: torch/distributions/categorical.py [hunk 2]
```diff
@@ -19,11 +19,13 @@ class Categorical(Distribution):
...
     probability of sampling the class at that index.
 
-    If :attr:`probs` is N-dimensional, the first N-1 dimensions are treated as a batch of 
+    If :attr:`probs` is N-dimensional, the first N-1 dimensions are treated as a batch of
     relative probability vectors.
 
     .. note:: :attr:`probs` must be non-negative, finite and have a non-zero sum,
               and it will be normalized to sum to 1 along the last dimension.
+              :attr:`logits` will be interpreted as unnormalized log probabilities
+              and can therefore be any real number.
```
[the remark this answers is not in the document]
Comment (neerajprad) [reply torch/distributions/categorical.py:28]: Good catch. I think that applies to `probs` attribute as well. I'll change this to argument instead for clarity.

File: torch/distributions/half_cauchy.py (hunk 1 above)
Comment (fehiepsi) [reply torch/distributions/half_cauchy.py:46]: Oops, you are right!

File: torch/distributions/relaxed_categorical.py [hunk 3]
```diff
@@ -30,8 +30,8 @@ class ExpRelaxedCategorical(Distribution):
     (Jang et al, 2017)
     """
     arg_constraints = {'probs': constraints.simplex,
-                       'logits': constraints.real}
-    support = constraints.real
+                       'logits': constraints.real_vector}
+    support = constraints.real_vector
```
Comment (neerajprad) [inline torch/distributions/relaxed_categorical.py:34]: cc. @fritzo.

Comment (fritzo) [reply torch/distributions/relaxed_categorical.py:34]: LGTM. I'm not sure whether `real_vector` is the optimal constraint, but I believe it is the tightest valid constraint that we currently implement. I think that's fine since most users will use the wrapper class with the exact `simplex` constraint.

**Review by fritzo — APPROVED** [review]

**Review by fehiepsi — APPROVED** [review]
